cli/cmd/upgrade.go

```diff
@@ -0,0 +1,207 @@
+		Get(k8s.ConfigConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		return nil, nil, err
+	}
+	uj := jsonpb.Unmarshaler{}
+	if err := uj.Unmarshal(strings.NewReader(configMap.Data["global"]), configs.Global); err != nil {
+		return nil, nil, err
+	}
+	if err := uj.Unmarshal(strings.NewReader(configMap.Data["proxy"]), configs.Proxy); err != nil {
+		return nil, nil, err
+	}
+
+	// Override the configs from the command-line flags.
+	options.overrideConfigs(configs)
+	j := jsonpb.Marshaler{EmitDefaults: true}
+	globalJSON, err := j.MarshalToString(configs.GetGlobal())
+	if err != nil {
+		return nil, nil, err
+	}
+	proxyJSON, err := j.MarshalToString(configs.GetProxy())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	values := &installValues{
+		// Container images:
+		ControllerImage: fmt.Sprintf("%s/controller:%s", options.dockerRegistry, configs.GetGlobal().GetVersion()),
+		WebImage:        fmt.Sprintf("%s/web:%s", options.dockerRegistry, configs.GetGlobal().GetVersion()),
+		GrafanaImage:    fmt.Sprintf("%s/grafana:%s", options.dockerRegistry, configs.GetGlobal().GetVersion()),
+		PrometheusImage: prometheusImage,
+		ImagePullPolicy: configs.Proxy.ProxyImage.PullPolicy,
+
+		// Kubernetes labels/annotations/resourcse:
+		CreatedByAnnotation:      k8s.CreatedByAnnotation,
+		CliVersion:               k8s.CreatedByAnnotationValue(),
+		ControllerComponentLabel: k8s.ControllerComponentLabel,
+		ProxyContainerName:       k8s.ProxyContainerName,
+		ProxyInjectAnnotation:    k8s.ProxyInjectAnnotation,
+		ProxyInjectDisabled:      k8s.ProxyInjectDisabled,
+
+		// Controller configuration:
+		Namespace:              controlPlaneNamespace,
+		UUID:                   configs.GetGlobal().GetInstallationUuid(),
+		ControllerLogLevel:     options.controllerLogLevel,
+		PrometheusLogLevel:     toPromLogLevel(options.controllerLogLevel),
+		ControllerReplicas:     controllerReplicas,
+		ControllerUID:          options.controllerUID,
+		EnableHA:               options.highAvailability,
+		EnableH2Upgrade:        !options.disableH2Upgrade,
+		NoInitContainer:        configs.GetGlobal().GetCniEnabled(),
+		ProxyAutoInjectEnabled: configs.GetGlobal().GetAutoinjectContext() != nil,
+
+		GlobalConfig: globalJSON,
+		ProxyConfig:  proxyJSON,
+	}
+
+	idctx := configs.GetGlobal().GetIdentityContext()
+	if idctx == nil {
+		// If we're upgrading from a version without identity, generate a new one.
+		i, err := newInstallIdentityOptionsWithDefaults().genValues()
+		if err != nil {
+			return nil, nil, err
+		}
+
+		values.Identity = i
+		return values, configs, nil
+	}
+
+	trustPEM := idctx.GetTrustAnchorsPem()
+	roots, err := tls.DecodePEMCertPool(trustPEM)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	secret, err := k.CoreV1().Secrets(controlPlaneNamespace).
+		Get(k8s.IdentityIssuerSecretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	keyPEM := string(secret.Data["key.pem"])
```

Comment:
templatize `key.pem` and `crt.pem` in `identity.yaml`

Reply:
i don't think it's worth templatizing but i'll const them at least